src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Emitter/CoreBindingHelpers.cs

```diff
@@ -633,7 +640,7 @@ private void EmitPopulationImplForArray(EnumerableSpec type)
 
                 // Create list and bind elements.
                 string tempIdentifier = GetIncrementalIdentifier(Identifier.temp);
```

Comment:
Seems to me that `InitializationKind.Declaration` and `ValueDefaulting.None` indicate the same scenario, when a new object is being declared. Maybe we could remove `ValueDefaulting`?

Reply:
The only place with `ValueDefaulting.CallSetter` is for complex types (objects) which also has  `InitializationKind.Declaration`. However, arrays and enumerables also have  `InitializationKind.Declaration` but should not have `ValueDefaulting.CallSetter`.

Reply:
Seems like something we can simplify by adding intermediary logic/methods before we get to the core binding methods. I wouldn't block on this though.

Generally I'm trying to tease out the decisions made in the core binding methods so that it's to understand (https://github.com/dotnet/runtime/pull/91717#discussion_r1320459650).

Reply:
The reason for the difference is because of the non-intuitive reflection semantics around not calling setter for objects added to arrays.

I could look at adding support for collections\enumerable; it would be a small breaking change for reflection, but that would mean modifying the reflection code.

Reply:
> I could look at adding support for collections\enumerable; it would be a small breaking change for reflection, but that would mean modifying the reflection code.

We should definitely avoid this, especially not for 8. The binder / user expectations are very sensitive to any breaking changes, as @tarekgh would probably echo.

Reply:
The approach you have now looks good to me.

Reply:
Yes, please try avoiding introducing any breaking change in for the reflection. Any slight change gets noticed and causes us servicing it. At least try to avoid it in the current stage of the project.